### PR TITLE
Add dispatcher script to base image

### PR DESCRIPTION
**Pull request by keaty**
## Purpose
Until now, entrypoint scripts have been run by custom wrapper scripts within each app. This leads to a lot of duplicate code that's time-consuming to update and easily falls out of date.

To standardize the practice a bit, this PR adds a dispatcher script to the ruby base image (`entrypoint_scripts/dispatcher.sh`) that takes:

1. Full custom script locations,
1. Or keywords like `use_dns_server` and runs the corresponding entrypoint script with the base docker command (e.g. `nodemon nerver.js`) as chained arguments.

An example of how that might look in a `docker-compose.yml`...
```
entrypoint:
    - docker-ssh-exec
    - /opt/entrypoint_scripts/dispatcher.sh
    - use_dns_server
    - wait_for_bundle_install
    - /bin/custom_repo_entrypoint_script.sh
    - bundle
    - --
```
As you can see, the keywords correspond to the name of the entrypoint script, minus `.sh`. The final `--` signifies the end of the dispatcher script arguments.

## Worth Noting
Of the entrypoint scripts that already exist, a few relied on having arguments passed to them by the wrapper scripts. This no longer is an option. External information required by entrypoint scripts is now accessed through environment variables. The scripts affected are:

* use_dns_server.sh (now uses $DNS_HOST)
* wait_for_bundle_install.sh (now uses $WAIT_FOR_BUNDLE)
* wait_for_node_install.sh (now uses $WAIT_FOR_NODE)
* service_health_checks/mysql.sh (now uses $WAIT_FOR_MYSQL)
* service_health_checks/elasticsearch.sh (now uses $ES_PORT)
* service_health_checks/redis.sh (now uses $REDIS_PORT)

## Launch Plan
Apps using entrypoint scripts should switch to using the new dispatcher script to run them, but will not break if they upgrade their base image without doing so.

Apps using entrypoint scripts *with custom parameters* will need to switch to environment variables *before* updating their base image.

## Gif (optional)
dispatcher

### CODE REVIEW

**Comment by jamiemccarthy** [comment]
If we eventually move to alpine, it only has `sh` (no `bash`) — does this script happen to work as-written on `#!/bin/sh`?

File: ruby/entrypoint_scripts/dispatcher.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+for arg in "$@"
+do
+  if [[ "$arg" == "--" ]]; then
+    shift
+    break
+  elif [[ $arg == *".sh"* ]]; then
+    eval $arg
+    shift
+  else
+    eval "/opt/entrypoint/$arg.sh" $BASE_COMMAND
```
Comment (thenonsequitur) [inline ruby/entrypoint_scripts/dispatcher.sh]: Is the `eval` necessary here?

**Comment by keaty** [comment]
Jamie, I've worked through what's required to translate the dispatcher script to `sh` and ...it's not pretty. It would make the script very "write-only". I'm leaning toward merging the PR as is, then initiating a longer conversation with the team regarding script maintainability vs small footprint base image.

**Comment by thenonsequitur** [comment]
The various entrypoint scripts themselves are also dependent upon bash for some of their functionality. I agree that switching away from bash is part of a larger conversation.

File: ruby/entrypoint_scripts/dispatcher.sh (hunk 1 above)
Comment (keaty) [reply ruby/entrypoint_scripts/dispatcher.sh]: Good point. Removed!

**Pull request merged**